OpenRA.Mods.Common/Traits/Render/SelectionDecorations.cs

```diff
@@ -165,6 +176,29 @@ IEnumerable<IRenderable> DrawPips(Actor self, int2 basePosition, PaletteReferenc
 				// Increment row
 				pipxyOffset = new int2(0, pipxyOffset.Y - (pipSize.Y + 1));
 			}
+
+			// Text pips.
+			foreach (var pips in textPipSources)
+			{
+				var pipType = pips.GetTextPipType(self);
+				var pipText = pips.GetTextPipText(self);
+				if (pipText == null)
+					continue;
+
+				if (pipxyOffset.X + pipSize.X >= width)
+					pipxyOffset = new int2(0, pipxyOffset.Y - pipSize.Y);
+
+				pipImages.PlayRepeating(PipStrings[(int)pipType]);
+				pipxyOffset += new int2(pipSize.X, 0);
+
+				yield return new UISpriteRenderable(pipImages.Image, self.CenterPosition, pipxyBase + pipxyOffset, 0, palette, 1f);
+
+				var textPos = pipxyBase + pipxyOffset + new int2(pipSize.X, 0);
+				yield return new UITextRenderable(font, self.CenterPosition, textPos, 0, Info.FontColor, " " + pipText);
```

Comment:
Modifying the string here undoes all the work to cache it in `AmmoPool`!
If you make the offset I requested above 2d then you can build the x offset in there.